drivers/flash/soc_flash_nrf_rram.c

```diff
@@ -0,0 +1,293 @@
+
+#define RRAM_WORD_SIZE 4
+
+#define PAGE_SIZE  DT_PROP(RRAM, erase_block_size)
+#define PAGE_COUNT ((RRAM_SIZE) / (PAGE_SIZE))
+
+#define ERASE_VALUE 0xFF
+
+static struct k_sem sem_lock;
+#define SYNC_INIT()   k_sem_init(&sem_lock, 1, 1)
+#define SYNC_LOCK()   k_sem_take(&sem_lock, K_FOREVER)
+#define SYNC_UNLOCK() k_sem_give(&sem_lock)
+
+#if CONFIG_NRF_RRAM_WRITE_BUFFER_SIZE > 0
+#define WRITE_BUFFER_ENABLE 1
+#define WRITE_BUFFER_SIZE   CONFIG_NRF_RRAM_WRITE_BUFFER_SIZE
+#define WRITE_LINE_SIZE     16 /* In bytes, one line is 128 bits. */
+#else
+#define WRITE_BUFFER_ENABLE 0
+#define WRITE_BUFFER_SIZE   0
+#define WRITE_LINE_SIZE     RRAM_WORD_SIZE
+#endif
+
+#ifndef CONFIG_SOC_FLASH_NRF_RADIO_SYNC_NONE
+#define FLASH_SLOT_WRITE 5000
+
+#if WRITE_BUFFER_ENABLE
+#define RRAM_MAX_WRITE_BUFFER (WRITE_BUFFER_SIZE * WRITE_LINE_SIZE)
+#else
+#define RRAM_MAX_WRITE_BUFFER (WRITE_LINE_SIZE * 32)
+#endif
+
+static int write_op(void *context); /* instance of flash_op_handler_t */
+static int write_synchronously(off_t addr, const void *data, size_t len);
+
+#endif /* !CONFIG_SOC_FLASH_NRF_RADIO_SYNC_NONE */
+
+static inline bool is_within_bounds(off_t addr, size_t len, off_t boundary_start,
+				    size_t boundary_size)
+{
+	return (addr >= boundary_start && (addr < (boundary_start + boundary_size)) &&
+		(len <= (boundary_start + boundary_size - addr)));
+}
+
+#if WRITE_BUFFER_ENABLE
+static void commit_changes(size_t len)
+{
+	if (nrf_rramc_empty_buffer_check(NRF_RRAMC)) {
+		/* The internal write-buffer has been committed to RRAM and is now empty. */
+		return;
+	}
+
+	if ((len % (WRITE_LINE_SIZE * WRITE_BUFFER_SIZE)) == 0) {
+		/* Our last operation was buffer size-aligned, so we're done. */
+		return;
+	}
+
+	nrf_rramc_task_trigger(NRF_RRAMC, NRF_RRAMC_TASK_COMMIT_WRITEBUF);
+
+	barrier_dmem_fence_full();
+}
+#endif
+
+static void rram_write(off_t addr, const void *data, size_t len)
+{
+	nrf_rramc_config_t config = {.mode_write = true, .write_buff_size = WRITE_BUFFER_SIZE};
+
+	nrf_rramc_config_set(NRF_RRAMC, &config);
+
+	if (data) {
+		memcpy((void *)addr, data, len);
+	} else {
+		memset((void *)addr, ERASE_VALUE, len);
+	}
+
+	barrier_dmem_fence_full(); /* Barrier following our last write. */
+
+#if WRITE_BUFFER_ENABLE
+	commit_changes(len);
+#endif
+
+	config.mode_write = false;
+	nrf_rramc_config_set(NRF_RRAMC, &config);
+}
+
+#ifndef CONFIG_SOC_FLASH_NRF_RADIO_SYNC_NONE
+static void shift_write_context(uint32_t shift, struct flash_context *w_ctx)
+{
+	w_ctx->flash_addr += shift;
+
+	/* NULL data_addr => erase emulation request*/
+	if (w_ctx->data_addr) {
+		w_ctx->data_addr += shift;
+	}
+
+	w_ctx->len -= shift;
+}
+
+static int write_op(void *context)
+{
+	struct flash_context *w_ctx = context;
+	size_t len;
+
+	uint32_t i = 0U;
+
+	if (w_ctx->enable_time_limit) {
+		nrf_flash_sync_get_timestamp_begin();
+	}
+
+	while (w_ctx->len > 0) {
+		len = (RRAM_MAX_WRITE_BUFFER < w_ctx->len) ? RRAM_MAX_WRITE_BUFFER : w_ctx->len;
```

Comment:
If we are in agreement with Flash API, so we write by `write_block_size`, then we either write by full buffers or `write_block_size`

Reply:
write_block_size is set to 1.